File: src/main/java/miragefairy2019/mod/lib/WeightedRandom.java
package miragefairy2019.mod.lib;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Random;
import java.util.function.BiPredicate;

import mirrg.boron.util.suppliterator.ISuppliterator;

public class WeightedRandom
{

	public static <T> Optional<T> getRandomItem(Random random, List<Item<T>> items)
	{
		if (items.size() <= 0) return Optional.empty();
		return Optional.of(getRandomItem(items, random.nextDouble() * getTotalWeight(items)));
	}

	public static double getTotalWeight(List<? extends Item<?>> items)
	{
		return items.stream()
			.mapToDouble(i -> i.weight)
			.sum();
	}

	private static <T> T getRandomItem(List<Item<T>> items, double w)
	{
		for (Item<T> item : items) {
			w -= item.weight;
			if (w < 0) return item.item;
		}
		return items.get(items.size() - 1).item;
	}

	public static final class Item<T>
	{

		public final double weight;
		public final T item;

		public Item(T item, double weight)
		{
			this.weight = weight;
			this.item = item;
		}

	}

	public static <T> List<Item<T>> unique(List<Item<T>> dropTable, BiPredicate<T, T> equals)
	{
		class A
		{

			private T t;

			public A(T t)
			{
				this.t = t;
			}

			@Override
			public int hashCode()
			{
				return 0;
			}

			@Override
			public boolean equals(Object obj)
			{
				if (this == obj) return true;
				if (obj == null) return false;
				if (getClass() != obj.getClass()) return false;
				A other = (A) obj;
				if (t == null) {
					if (other.t != null) return false;
				} else if (!equals.test(t, other.t)) return false;
				return true;
			}

		}

		Map<A, Double> map = new HashMap<>();

		for (Item<T> item : dropTable) {
			map.put(new A(item.item), map.getOrDefault(new A(item.item), 0.0) + item.weight);
		}

		return ISuppliterator.ofIterable(map.entrySet())
			.map(e -> new Item<>(e.getKey().t, e.getValue()))
			.toList();
	}
}

File: src/main/java/miragefairy2019/libkt/WeightedRandom.kt
// TODO change
package miragefairy2019.mod.lib

import java.util.Random
import java.util.function.BiPredicate

fun <T : Any> List<WeightedRandom.WeightedItem<T>>.getRandomItem(random: Random) = getItem(random.nextDouble())

/** @param d 0以上1未満の値 */
fun <T : Any> List<WeightedRandom.WeightedItem<T>>.getItem(d: Double): T? {
    if (isEmpty()) return null

    var w = d * totalWeight // 0 <= w < totalWeight
    this.forEach { item ->
        w -= item.weight
        if (w < 0) return item.item
    }
    return this.last().item
}

val <T : Any> List<WeightedRandom.WeightedItem<T>>.totalWeight get() = sumByDouble { it.weight }

// TODO flat
object WeightedRandom {

    // TODO -> Pair
    class WeightedItem<T : Any>(val item: T, val weight: Double)

    // TODO receiver
    fun <T : Any> unique(dropTable: List<WeightedItem<T>>, equals: BiPredicate<T, T>): List<WeightedItem<T>> {
        class Slot(val item: T) {
            override fun hashCode() = 0

            override fun equals(other: Any?): Boolean {
                if (this === other) return true // 相手が自分自身なら一致
                if (other == null) return false // 相手が無なら不一致

                // 型チェック
                if (javaClass != other.javaClass) return false
                @Suppress("UNCHECKED_CAST")
                other as Slot

                return equals.test(item, other.item)
            }
        }

        val map = mutableMapOf<Slot, Double>()
        dropTable.forEach { item ->
            map[Slot(item.item)] = (map[Slot(item.item)] ?: 0.0) + item.weight
        }
        return map.entries.map { WeightedItem<T>(it.key.item, it.value) }
    }
}
